lib/bolt/shell/powershell.rb

```diff
@@ -0,0 +1,263 @@
+                      path, args = *process_from_extension(remote_path)
+                      args += escape_arguments(arguments)
+                      execute_process(path, args)
+                    end
+          output = execute(command)
+          Bolt::Result.for_command(target,
+                                   output.stdout.string,
+                                   output.stderr.string,
+                                   output.exit_code,
+                                   'script', script)
+        end
+      end
+
+      def run_task(task, arguments, _options = {})
+        implementation = select_implementation(target, task)
+        executable = implementation['path']
+        input_method = implementation['input_method']
+        extra_files = implementation['files']
+        input_method ||= powershell_file?(executable) ? 'powershell' : 'both'
+
+        # unpack any Sensitive data
+        arguments = unwrap_sensitive_args(arguments)
+        with_remote_tempdir do |dir|
+          if extra_files.empty?
+            task_dir = dir
+          else
+            # TODO: optimize upload of directories
+            arguments['_installdir'] = dir
+            task_dir = File.join(dir, task.tasks_dir)
+            mkdirs([task_dir] + extra_files.map { |file| File.join(dir, File.dirname(file['name'])) })
+            extra_files.each do |file|
+              conn.copy_file(file['path'], File.join(dir, file['name']))
+            end
+          end
+
+          remote_task_path = write_executable(task_dir, executable)
+
+          if Bolt::Task::STDIN_METHODS.include?(input_method)
+            stdin = JSON.dump(arguments)
+          end
+
+          command = if powershell_file?(remote_task_path) && stdin.nil?
+                      run_ps_task(remote_task_path, arguments, input_method)
+                    else
+                      if (interpreter = select_interpreter(remote_task_path, target.options['interpreters']))
+                        path = interpreter
+                        args = [remote_task_path]
+                      else
+                        path, args = *process_from_extension(remote_task_path)
+                      end
+                      execute_process(path, args, stdin)
+                    end
+
+          env_assignments = if Bolt::Task::ENVIRONMENT_METHODS.include?(input_method)
+                              envify_params(arguments).map do |(arg, val)|
+                                set_env(arg, val)
+                              end
+                            else
+                              []
+                            end
+
+          output = execute([Snippets.shell_init, *env_assignments, command].join("\n"))
+
+          Bolt::Result.for_task(target, output.stdout.string,
+                                output.stderr.string,
+                                output.exit_code,
+                                task.name)
+        end
+      end
+
+      def execute(command)
```

Comment:
This seems like it should be the `execute` method that's defined in `transport/winrm/connection.rb` - that definition of execute doesn't look to be called, but should move here to preserve logging

Reply:
I'm not sure I follow. This `execute` method calls `execute` on the connection.

Reply:
Ahhh I'm not sure what I thought this was doing but it makes sense.